README.md
# Dart package for the Azure Cosmos Database

<div>
<a href="https://pub.dev/packages/dart_azure_cosmosdb"><img src="https://img.shields.io/pub/v/dart_azure_cosmosdb.svg" alt="Pub"></a>
<img alt="GitHub Workflow Status (branch)" src="https://img.shields.io/github/actions/workflow/status/Pat992/dart_azure_cosmosdb/dart.yml?branch=main">
<a href="https://github.com/Pat992/dart_azure_cosmosdb"><img src="https://img.shields.io/github/stars/Pat992/dart_azure_cosmosdb?style=flat&logo=github&colorB=deeppink&label=stars" alt="Star on Github"></a>
<a href="https://opensource.org/licenses/BSD-3-Clause"><img src="https://img.shields.io/static/v1?label=license&message=BSD 3-Clause&color=blue" alt="License: BSD 3-Clause"></a>
</div>

## Table of content

- [1 Features](#1-features)
- [2 Getting started](#2-getting-started)
  - [2.1 Android settings](#21-android-settings)
  - [2.2 iOS settings](#21-ios-settings)
  - [2.3 macOS settings](#23-macos-settings)
  - [2.4 Windows settings](#24-windows-settings)
  - [2.5 Linux settings](#25-linux-settings)
- [3 Usage](#3-usage)
  - [3.1 CosmosDB instance creation and configuration](#31-cosmosdb-instance-creation-and-configuration)
  - [3.2 Databases](#32-databases)
    - [3.2.1 List Databases](#321-list-databases)
    - [3.2.2 Get a Database](#321-get-a-database)
    - [3.2.3 Create a Database](#322-create-a-database)
    - [3.2.4 Delete a Database](#323-delete-a-database)
    - [3.2.5 The CosmosDatabase model](#325-the-cosmosdatabase-model)
  - [3.3 Collections](#53-collections)
    - [3.3.1 List Collections](#331-list-collections)
    - [3.3.2 Get a Collection](#332-get-a-collection)
    - [3.3.3 Create a Collection](#333-create-a-collection)
    - [3.3.4 Delete a Collection](#334-delete-a-collection)
    - [3.3.5 The Collection model](#335-the-collection-model)
  - [3.4 Documents](#34-documents)
    - [3.4.1 List Documents](#341-list-documents)
    - [3.4.2 Get a Document](#342-get-a-document)
    - [3.4.3 Create a Document](#343-create-a-document)
    - [3.4.4 Replace a Document](#344-replace-a-document)
    - [3.4.5 Delete a Document](#345-delete-a-document)
    - [3.4.6 Query documents](#346-query-documents)
    - [3.4.7 The CosmosDocument model](#347-the-cosmosdocument-model)
  - [3.5 Stored Procedures](#35-stored-procedures)
    - [3.5.1 List Stored Procedures](#351-list-stored-procedures)
    - [3.5.2 Create a Stored Procedure](#352-create-a-stored-procedure)
    - [3.5.3 Replace a Stored Procedure](#353-replace-a-stored-procedure)
    - [3.5.4 Delete a Stored Procedure](#354-delete-a-stored-procedure)
    - [3.5.5 Execute a Stored Procedure](#355-execute-a-stored-procedure)
    - [3.5.6 The Stored Procedure model](#356-the-stored-procedure-model)
  - [3.6 User Defined Functions](#36-user-defined-functions)
    - [3.6.1 List User Defined Functions](#361-list-user-defined-functions)
    - [3.6.2 Create a User Defined Function](#362-create-a-user-defined-function)
    - [3.6.3 Replace a User Defined Function](#363-replace-a-user-defined-function)
    - [3.6.4 Delete a User Defined Function](#364-delete-a-user-defined-function)
    - [3.6.5 The User Defined Function model](#365-the-user-defined-function-model)
  - [3.7 Triggers](#37-triggers)
    - [3.7.1 List Triggers](#371-list-triggers)
    - [3.7.2 Create a Trigger](#372-create-a-trigger)
    - [3.7.3 Replace a Trigger](#373-replace-a-trigger)
    - [3.7.4 Delete a Trigger](#374-delete-a-trigger)
    - [3.7.5 The Trigger model](#375-the-trigger-model)
  - [3.8 Users](#38-users)
    - [3.8.1 List Users](#381-list-users)
    - [3.8.2 Get a User](#382-get-a-user)
    - [3.8.3 Create a User](#383-create-a-user)
    - [3.8.4 Replace a User](#384-replace-a-user)
    - [3.8.5 Delete a User](#385-delete-a-user)
    - [3.8.6 The Trigger model](#386-the-trigger-model)
  - [3.9 Permissions](#39-permissions)
    - [3.9.1 List Permissions](#391-list-permissions)
    - [3.9.2 Get a Permission](#392-get-a-permission)
    - [3.9.3 Create a Permission](#393-create-a-permission)
    - [3.9.4 Replace a Permission](#394-replace-a-permission)
    - [3.9.5 Delete a Permission](#395-delete-a-permission)
    - [3.9.6 The Permission model](#396-the-permission-model)
- [4 Where to go from here](#4-where-to-go-from-here)
- [5 Bugs and issues](#5-bugs-and-issues)

## 1 Features
**dart_azure_cosmosdb** offers interaction with the Azure CosmosDB SQL-API.
For more information, check out the **[Microsoft CosmosDB documentation](https://learn.microsoft.com/en-us/rest/api/cosmos-db/)**.
## 2 Getting started

### 2.1 Android settings
This package uses the HTTP-package and needs access to the internet, networking needs to be enabled in the `AndroidManifest.xml`.
```xml
<manifest xmlns:android...>
    ...
    <uses-permission android:name="android.permission.INTERNET" />
    <application ...
</manifest>
```

### 2.2 iOS settings
No steps are necessary for iOS.

### 2.3 macOS settings
This package uses the HTTP-package and needs access to the internet, networking needs to be enabled in the `.entitlements`.
```xml
<key>com.apple.security.network.client</key>
<true/>
```

### 2.4 Windows settings
No steps are necessary for Windows.

### 2.5 Linux settings
No steps are necessary for Linux.

## 3 Usage

### 3.1 CosmosDB instance creation and configuration
First, you need to create an instance of the CosmosDB-class, as follows:
```dart
final cosmosDb = CosmosDb(
    // REQUIRED -> URI to the Azure Cosmos Database.
    connectionUri: 'https://<COSMOS_DB_NAME>.documents.azure.com:443/',
    // REQUIRED -> Primary key to the Cosmos Database.
    primaryKey: '<COSMOS_DB_KEY>',
    // OPTIONAL -> Authorization type of the provided token.
    // Default: master
    authorizationType: 'master',
    // OPTIONAL -> Version of the provided authorization.
    // Default: 1.0
    authorizationVersion: '1.0',
    // OPTIONAL -> The supported REST API versions of the Azure Cosmos DB service.
    // Default: 2018-12-31
    xmsVersion: '2018-12-31',
);
```

### 3.2 Databases
Each Azure Cosmos DB account supports multiple independently named databases. A database is a logical namespace for collections, users, and permissions.

#### 3.2.1 List Databases
List the databases under the database account, performs a GET on the databases resource.

The database-list operation returns a `Future<CosmosDatabaseList>` with a list of  `CosmosDatabase` or an error object.
See [3.2.5 The CosmosDatabase model](#325-the-cosmosdatabase-model) for information on the returned model and its variables.
```dart
final CosmosDatabaseList databaseList = await cosmosDb.database.list();

if(databaseList.error.isEmpty) {
    // If the error is empty, continue....
}
```

#### 3.2.2 Get a Database
Retrieve a database resource, performs a GET on the database resource.

The database-get operation returns a `Future<CosmosDatabase>`.
See [3.2.5 The CosmosDatabase model](#325-the-cosmosdatabase-model) for information on the returned model and its variables.
```dart
final CosmosDatabase database = await cosmosDb.database.get(
    // REQUIRED -> User generated name/id of the database.
    dbId: '<DB_ID>',
);

if (database.error.isEmpty) {
    // If the error is empty, continue....
}
```

#### 3.2.3 Create a Database
The create-operation creates a new database in the database account.

The database-create operation returns a `Future<CosmosDatabase>`.
See [3.2.5 The CosmosDatabase model](#325-the-cosmosdatabase-model) for information on the returned model and its variables.
```dart
final CosmosDatabase database = await cosmosDb.database.create(
    // REQUIRED -> User generated name/id of the database to be created.
    dbId: '<DB_ID>',
);

if (database.error.isEmpty) {
    // If the error is empty, continue....
}
```

#### 3.2.4 Delete a Database
The delete-operation deletes an existing database in the database account.

The database-delete operation returns a `Future<CosmosDatabase>`.
See [3.2.5 The CosmosDatabase model](#325-the-cosmosdatabase-model) for information on the returned model and its variables.
```dart
final CosmosDatabase database = await cosmosDb.database.delete(
    // REQUIRED -> User generated name/id of the database to be deleted.
    dbId: '<DB_ID>',
);

if (database.error.isEmpty) {
    // If the error is empty, continue....
}
```

#### 3.2.5 The CosmosDatabase model
##### `CosmosDatabase`

**Methods**
`CosmosDatabase.toMap()` -> Transforms the model to a Map for the CosmosDB API.
`CosmosDatabase.toMap(Map<String, dynamic> map)` -> Transforms a Map from the CosmosDB API to the model.

**Variables**
<table>
    <thead>
        <tr>
            <th>Name</th>
            <th>Azure API key name</th>
            <th>Example</th>
            <th>Description</th>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td><code>id</code></td>
            <td><code>id</code></td>
            <td>myId1</td>
            <td>The unique user generated name for the resource, max 255 chars.</td>
        </tr>
        <tr>
            <td><code>rid</code></td>
            <td><code>_rid</code></td>
            <td>qicAAA==</td>
            <td>System generated property. The resource ID (_rid) is a unique identifier. It is used internally for placement of and navigation to the resource.</td>
        </tr>
        <tr>
            <td><code>ts</code></td>
            <td><code>_ts</code></td>
            <td>1446192371</td>
            <td>System generated property. It specifies the last updated timestamp of the resource. The value is a timestamp.</td>
        </tr>
        <tr>
            <td><code>self</code></td>
            <td><code>_self</code></td>
            <td>dbs\/qicAAA==\/</td>
            <td>System generated property. It is the unique addressable URI for the resource.</td>
        </tr>
        <tr>
            <td><code>etag</code></td>
            <td><code>_etag</code></td>
            <td>00001800-...</td>
            <td>System generated property representing the resource etag required for optimistic concurrency control.</td>
        </tr>
        <tr>
            <td><code>error</code></td>
            <td><code>code</code>,<br><code>message</code></td>
            <td>{<br>"code": "...",<br>"message": "..."<br>}</td>
            <td>An empty Map in case of a successful transaction, else a map with a code and a message.</td>
        </tr>
        <tr>
            <td><code>colls</code></td>
            <td><code>_colls</code></td>
            <td>colls/</td>
            <td>It is a system generated property that specifies the addressable path of the collections resource.</td>
        </tr>
        <tr>
            <td><code>users</code></td>
            <td><code>_users</code></td>
            <td>users/</td>
            <td>It is a system generated property that specifies the addressable path of the users resource.</td>
        </tr>
    </tbody>
</table>

##### `CosmosDatabaseList`
**Methods**
`CosmosDatabaseList.toMap()` -> Transforms the model to a Map for the CosmosDB API.
`CosmosDatabaseList.toMap(Map<String, dynamic> map)` -> Transforms a Map from the CosmosDB API to the model.

**Variables**
<table>
    <thead>
        <tr>
            <th>Name</th>
            <th>Azure API key name</th>
            <th>Example</th>
            <th>Description</th>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td><code>rid</code></td>
            <td><code>_rid</code></td>
            <td>qicAAA==</td>
            <td>System generated property. The resource ID (_rid) is a unique identifier. It is used internally for placement of and navigation to the resource.</td>
        </tr>
        <tr>
            <td><code>count</code></td>
            <td><code>_count</code></td>
            <td>3</td>
            <td>This property shows the number of resources returned as part of list operation.</td>
        </tr>
        <tr>
            <td><code>error</code></td>
            <td><code>code</code>,<br><code>message</code></td>
            <td>{<br>"code": "...",<br>"message": "..."<br>}</td>
            <td>An empty Map in case of a successful transaction, else a map with a code and a message.</td>
        </tr>
        <tr>
            <td><code>databases</code></td>
            <td><code>Databases</code></td>
            <td>[]</td>
            <td>A list of CosmosDB databases. <code>List&lt;CosmosDatabase&gt;</code></td>
        </tr>
    </tbody>
</table>

### 3.3 Collections
A collection is a container of JSON documents and associated JavaScript application logic, i.e. stored procedures, triggers and user-defined functions.

#### 3.3.1 List Collections
Performing a GET on the collections resource of a particular database, i.e. the colls URI path, returns a list of the collections in the database.

The collection-list operation returns a `Future<CollectionList>` with a list of  `Collection` or an error object.
See [3.3.5 The Collection model](#335-the-collection-model) for information on the returned model and its variables.
```dart
final CollectionList collectionList = await cosmosDb.collection.list(
    // REQUIRED -> User generated name/id of the database in which the collections are located.
    dbId: '<DB_ID>',
);

if (collectionList.error.isEmpty) {
    // If the error is empty, continue....
}
```

#### 3.3.2 Get a Collection
Performing a GET on a specific collection resource retrieves the properties for the collection.

The collection-get operation returns a `Future<Collection>`.
See [3.3.5 The Collection model](#335-the-collection-model) for information on the returned model and its variables.
```dart
final Collection collection = await cosmosDb.collection.get(
    // REQUIRED -> User generated name/id of the database in which the collection is located.
    dbId: '<DB_ID>',
    // REQUIRED -> Name of the existing collection.
    collectionId: '<COLLECTION_ID>',
);

if (collection.error.isEmpty) {
    // If the error is empty, continue....
}
```

#### 3.3.3 Create a Collection
The create-operation creates a new collection in a database.

The collection-create operation returns a `Future<Collection>`.
See [3.3.5 The Collection model](#335-the-collection-model) for information on the returned model and its variables.
```dart
final Collection collection = await cosmosDb.collection.create(
    // REQUIRED -> User generated name/id of the database in which the collections are saved.
    dbId: '<DB_ID>',
    // REQUIRED -> Name of the collection to be created.
    collectionId: '<COLLECTION_ID>',
    // REQUIRED -> The partition key to be used for partitioning data into multiple partitions.
    partitionKey: '/id',
    // REQUIRED -> Set this value to 2, to use a large partition key, else set it to 1.
    version: 1,
);

if (collection.error.isEmpty) {
    // If the error is empty, continue....
}
```

#### 3.3.4 Delete a Collection
The delete-operation deletes an existing collection in a database.

The collection-delete operation returns a `Future<Collection>`.
See [3.3.5 The Collection model](#335-the-collection-model) for information on the returned model and its variables.
```dart
final Collection collection = await cosmosDb.collection.delete(
    // REQUIRED -> User generated name/id of the database in which the collection is located.
    dbId: '<DB_ID>', 
    // REQUIRED -> Name of the collection to be deleted.
    collectionId: '<COLLECTION_ID>',
);

if (collection.error.isEmpty) {
    // If the error is empty, continue....
}
```

#### 3.3.5 The Collection model
##### `Collection`
**Methods**
`Collection.toMap()` -> Transforms the model to a Map for the CosmosDB API.
`Collection.toMap(Map<String, dynamic> map)` -> Transforms a Map from the CosmosDB API to the model.

**Variables**
<table>
    <thead>
        <tr>
            <th>Name</th>
            <th>Azure API key name</th>
            <th>Example</th>
            <th>Description</th>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td><code>id</code></td>
            <td><code>id</code></td>
            <td>myId1</td>
            <td>The unique user generated name for the resource, max 255 chars.</td>
        </tr>
        <tr>
            <td><code>rid</code></td>
            <td><code>_rid</code></td>
            <td>qicAAA==</td>
            <td>System generated property. The resource ID (_rid) is a unique identifier. It is used internally for placement of and navigation to the resource.</td>
        </tr>
        <tr>
            <td><code>ts</code></td>
            <td><code>_ts</code></td>
            <td>1446192371</td>
            <td>System generated property. It specifies the last updated timestamp of the resource. The value is a timestamp.</td>
        </tr>
        <tr>
            <td><code>self</code></td>
            <td><code>_self</code></td>
            <td>dbs\/qicAAA==\/</td>
            <td>System generated property. It is the unique addressable URI for the resource.</td>
        </tr>
        <tr>
            <td><code>etag</code></td>
            <td><code>_etag</code></td>
            <td>00001800-...</td>
            <td>System generated property representing the resource etag required for optimistic concurrency control.</td>
        </tr>
        <tr>
            <td><code>error</code></td>
            <td><code>code</code>,<br><code>message</code></td>
            <td>{<br>"code": "...",<br>"message": "..."<br>}</td>
            <td>An empty Map in case of a successful transaction, else a map with a code and a message.</td>
        </tr>
        <tr>
            <td><code>docs</code></td>
            <td><code>_docs</code></td>
            <td>docs/</td>
            <td>It is a system generated property that specifies the addressable path of the documents resource.</td>
        </tr>
        <tr>
            <td><code>sprocs</code></td>
            <td><code>_sprocs</code></td>
            <td>sprocs/</td>
            <td>It is a system generated property that specifies the addressable path of the stored procedures (sprocs) resource.</td>
        </tr>
        <tr>
            <td><code>triggers</code></td>
            <td><code>_triggers</code></td>
            <td>triggers/</td>
            <td>It is a system generated property that specifies the addressable path of the triggers resource.</td>
        </tr>
        <tr>
            <td><code>udfs</code></td>
            <td><code>_udfs</code></td>
            <td>udfs/</td>
            <td>It is a system generated property that specifies the addressable path of the user-defined functions (udfs) resource.</td>
        </tr>
        <tr>
            <td><code>conflicts</code></td>
            <td><code>_conflicts</code></td>
            <td>conflicts/</td>
            <td>It is a system generated property that specifies the addressable path of the conflicts resource.</td>
        </tr>
        <tr>
            <td><code>partitionKey</code></td>
            <td><code>partitionKey</code></td>
            <td>{<br>"paths": [ "/AccountNumber" ],<br>"kind": "Hash",<br>"Version": 2<br>}</td>
            <td>This value is used to configure the partition key to be used for partitioning data into multiple partitions.</td>
        </tr>
        <tr>
            <td><code>indexingPolicy</code></td>
            <td><code>indexingPolicy</code></td>
            <td>[{<br>"path": "/*",<br>"indexes": [{ "dataType": "String",<br>"precision": -1,<br> "kind": "Range" }]<br>}]</td>
            <td>This value is used to configure indexing policy. By default, the indexing is automatic for all document paths within the collection.</td>
        </tr>
        <tr>
            <td><code>uniqueKeyPolicy</code></td>
            <td><code>uniqueKeyPolicy</code></td>
            <td>[]</td>
            <td>With unique keys, you make sure that one or more values within a logical partition is unique. By default, the unique key policy is an empty array or null as an object.</td>
        </tr>
        <tr>
            <td><code>geospatialConfig</code></td>
            <td><code>geospatialConfig</code></td>
            <td>{<br>"type": "Geography"<br>}</td>
            <td>The Geospatial Configuration specifies how the spatial data will be indexed. Specify one Geospatial Configuration per container: geography or geometry.</td>
        </tr>
    </tbody>
</table>

##### `CollectionList`
**Methods**
`CollectionList.toMap()` -> Transforms the model to a Map for the CosmosDB API.
`CollectionList.toMap(Map<String, dynamic> map)` -> Transforms a Map from the CosmosDB API to the model.

**Variables**
<table>
    <thead>
        <tr>
            <th>Name</th>
            <th>Azure API key name</th>
            <th>Example</th>
            <th>Description</th>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td><code>rid</code></td>
            <td><code>_rid</code></td>
            <td>qicAAA==</td>
            <td>System generated property. The resource ID (_rid) is a unique identifier. It is used internally for placement of and navigation to the resource.</td>
        </tr>
        <tr>
            <td><code>count</code></td>
            <td><code>_count</code></td>
            <td>3</td>
            <td>This property shows the number of resources returned as part of list operation.</td>
        </tr>
        <tr>
            <td><code>error</code></td>
            <td><code>code</code>,<br><code>message</code></td>
            <td>{<br>"code": "...",<br>"message": "..."<br>}</td>
            <td>An empty Map in case of a successful transaction, else a map with a code and a message.</td>
        </tr>
        <tr>
            <td><code>collections</code></td>
            <td><code>DocumentCollections</code></td>
            <td>[]</td>
            <td>A list of CosmosDB collections. <code>List&lt;Collection&gt;</code></td>
        </tr>
    </tbody>
</table>

### 3.4 Documents
A document consists of user-defined content in JSON format. Aside from the required ID property, users can define any arbitrary elements, structures, and hierarchies for the content. The ID element is a unique string that is user-settable and must not exceed 255 characters. You can insert, replace, delete, read, enumerate, and query arbitrary JSON documents in a collection.

#### 3.4.1 List Documents
Performing a GET on the documents resource of a particular collection, i.e. the docs URI path, returns a list of documents under the collection.

The document-list operation returns a `Future<CosmosDocumentList>` with a list of  `CosmosDocument` or an error object.
See [3.4.6 The CosmosDocument model](#346-the-cosmosdocument-model) for information on the returned model and its variables.
```dart
final CosmosDocumentList documentList = await cosmosDb.document.list(
    // REQUIRED -> User generated name/id of the database.
    dbId: '<DB_ID>',
    // REQUIRED -> Id/name of the collection.
    collectionId: '<COLLECTION_ID>',
);

if (documentList.error.isEmpty) {
    // If the error is empty, continue....
}
```

#### 3.4.2 Get a Document
The get-operation retrieves a document by its partition key and document key.

The collection-get operation returns a `Future<CosmosDocument>`.
See [3.4.6 The CosmosDocument model](#346-the-cosmosdocument-model) for information on the returned model and its variables.
```dart
final CosmosDocument document = await cosmosDb.document.get(
    // REQUIRED -> User generated name/id of the database.
    dbId: '<DB_ID>',
    // REQUIRED -> Id/name of the collection.
    collectionId: '<COLLECTION_ID>',
    // REQUIRED -> Id of the document to be retrieved.
    documentId: '<DOCUMENT_ID>',
    // REQUIRED -> Partition key value for the document to be received.
    partitionKey: '<DOCUMENT_PARTITION_KEY>',
);

if (document.error.isEmpty) {
    // If the error is empty, continue....
}
```

#### 3.4.3 Create a Document
The create-operation creates a new document in a collection.

The collection-create operation returns a `Future<CosmosDocument>`.
See [3.4.6 The CosmosDocument model](#346-the-cosmosdocument-model) for information on the returned model and its variables.
```dart

// Example: Create an instance of a Person to write to the document.
final Person person = Person(
    id: '01',
    name: 'Jane', 
    familyName: 'Doe', 
    birthDate: '01.01.1970',
);

final CosmosDocument document = await cosmosDb.document.create(
    // REQUIRED -> User generated name/id of the database.
    dbId: '<DB_ID>',
    // REQUIRED -> Id/name of the collection.
    collectionId: '<COLLECTION_ID>',
    // REQUIRED -> Partition key value for the document to be created.
    // In this example the partition-key /id is used.
    partitionKey: person.id,
    // REQUIRED -> Any map with Strings as keys and dynamic values that will be added to the document.
    // In this example the Person-instance is transformed into a map.
    body: person.toMap(),
);

if (document.error.isEmpty) {
    // If the error is empty, continue....
}
```

#### 3.4.4 Replace a Document
The replace-operation replaces the entire content of a document.

The collection-replace operation returns a `Future<CosmosDocument>`.
See [3.4.6 The CosmosDocument model](#346-the-cosmosdocument-model) for information on the returned model and its variables.
```dart

CHANGELOG.md
## 1.0.0
- Fixed Shield for build result
- Updated packages http and crypto

## 0.9.0+1
- Documentation updates for the badges
- pubspec.yaml update to follow the guidelines

## 0.9.0
- Initial version.

lib/src/dart_azure_cosmosdb_base.dart
// Copyright 2023 Patrick Hettich. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be found in the LICENSE file.
import 'package:dart_azure_cosmosdb/src/core/auth_util.dart';
import 'package:dart_azure_cosmosdb/src/core/date_util.dart';
import 'package:dart_azure_cosmosdb/src/domain/repositories/i_repository_barrel.dart';
import 'package:dart_azure_cosmosdb/src/infrastructure/datasources/datasources_barrel.dart';
import 'package:dart_azure_cosmosdb/src/infrastructure/repositories/repository_barrel.dart';
import 'package:http/http.dart' as http;

class CosmosDb {
  /// **Description:** Authorization type of the provided token.
  ///
  /// **Default value:** 'master'
  final String authorizationType;

  /// **Description:** Version of the provided authorization.
  ///
  /// **Default value:** '1.0'
  final String authorizationVersion;

  /// **Description:** Base URI to the Azure Cosmos Database.
  ///
  /// **Example:** 'https://<COSMOS_DB_NAME>.documents.azure.com:443/'
  final String connectionUri;

  /// **Description:** The supported REST API versions of the Azure Cosmos DB service.
  ///
  /// **Default value:** '2018-12-31'
  final String xmsVersion;

  /// **Description:** Primary key to the Cosmos Database.
  final String primaryKey;

  /// **Description:** The collection-datasource, used by the collection-repository.
  late final ICollectionDatasource _collectionDatasource;

  /// **Description:** The database-datasource, used by the database-repository.
  late final IDatabaseDatasource _databaseDatasource;

  /// **Description:** The document-datasource, used by the document-repository.
  late final IDocumentDatasource _documentDatasource;

  /// **Description:** The permission-datasource, used by the permission-repository.
  late final IPermissionDatasource _permissionDatasource;

  /// **Description:** The stored-procedure-datasource, used by the stored-procedure-repository.
  late final IStoredProcedureDatasource _storedProcedureDatasource;

  /// **Description:** The trigger-datasource, used by the trigger-repository.
  late final ITriggerDatasource _triggerDatasource;

  /// **Description:** The user-datasource, used by the user-repository.
  late final IUserDatasource _userDatasource;

  /// **Description:** The user-defined-function-datasource, used by the user-defined-function-repository.
  late final IUserDefinedFunctionDatasource _userDefinedFunctionDatasource;

  /// **Description:** The http-client used in the datasources
  final _client = http.Client();

  /// **Description:** The auth-util, used for creating the http-headers.
  final IAuthUtil _authUtil = AuthUtil();

  /// **Description:** The date-util, used for creating and building the current date for the http-headers.
  final IDateUtil _dateUtil = DateUtil();

  /// **Description:** The REST API supports basic CRUD operations on the resources under a database account.
  /// A collection is a container of JSON documents and associated JavaScript application logic, i.e. stored procedures, triggers and user-defined functions.
  /// This topic outlines the REST operations used to manage document collections.
  late final ICollectionRepository collection;

  /// **Description:** A database is a logical namespace for collections, users, and permissions.
  late final IDatabaseRepository database;

  /// **Description:** The document resource is represented by docs in the Azure Cosmos DB resource model.
  /// A document consists of user-defined content in JSON format.
  /// Aside from the required ID property, users can define any arbitrary elements, structures, and hierarchies for the content.
  /// The ID element is a unique string that is user-settable and must not exceed 255 characters. You can insert, replace, delete, read, enumerate, and query arbitrary JSON documents in a collection.
  late final IDocumentRepository document;

  /// **Description:** A permission is an authorization token associated with a user for authorized access to a specific resource.
  /// It is used to manage access to collections, documents, attachments, stored procedures, triggers, and user-defined functions for a particular user.
  late final IPermissionRepository permission;

  /// **Description:** A stored procedure is a piece of application logic written in JavaScript that is registered and executed against a collection as a single transaction.
  /// In Azure Cosmos DB, JavaScript is hosted in the same memory space as the database.
  /// Hence, requests made within stored procedures execute in the same scope of a database session.
  /// This process enables Azure Cosmos DB to guarantee ACID for all operations that are part of a single stored procedure.
  late final IStoredProcedureRepository storedProcedure;

  /// **Description:** Triggers are pieces of application logic that can be executed before (pre-triggers) and after (post-triggers) creation, deletion, and replacement of a document.
  /// Triggers are written in JavaScript.
  /// Both pre and post triggers do no take parameters.
  /// Like stored procedures, triggers live within the confines of a collection, thus confining the application logic to the collection.
  late final ITriggerRepository trigger;

  /// **Description:** A user is a logical namespace for scoping permissions on collections, documents, attachment, stored procedures, triggers, and user-defined functions.
  /// The user construct lives under a database resource and thus cannot cross the database boundary it is under.
  /// The user resource is represented by users in the Azure Cosmos DB REST resource model.
  /// It is a system resource, and thus contains a fixed schema.
  late final IUserRepository user;

  /// **Description:** A user-defined function (UDF) is a side effect free piece of application logic written in JavaScript.
  /// It allows developers to construct a query operator, thus extending the core of the Cosmos DB query language.
  /// Like stored procedures, UDFs live in the confines of a collection, thus confining the application logic to the collection.
  late final IUserDefinedFunctionRepository userDefinedFunctions;

  /// **Description:** Creates a [CosmosDb] instance.
  ///
  /// **connectionUri:** The URI to the Azure Cosmos Database,
  /// for example: https://<COSMOS_DB_NAME>.documents.azure.com:443/
  ///
  /// [Type:] String *required*
  ///
  /// **primaryKey:** Primary key to the Cosmos Database.
  ///
  /// [Type:] String *required*
  ///
  /// **authorizationType:** Authorization type of the provided token.
  ///
  /// [Type:] String *optional*
  ///
  /// [Default value:] 'master'
  ///
  /// **authorizationVersion:** Version of the provided authorization.
  ///
  /// [Type:] String *optional*
  ///
  /// [Default value:] '1.0'
  ///
  /// **xmsVersion:** The supported REST API versions of the Azure Cosmos DB service.
  ///
  /// [Type:] String *optional*
  ///
  /// [Default value:] '2018-12-31'
  CosmosDb({
    required this.connectionUri,
    required this.primaryKey,
    this.authorizationType = 'master',
    this.authorizationVersion = '1.0',
    this.xmsVersion = '2018-12-31',
  }) {
    _collectionDatasource = CollectionDatasource(
      client: _client,
      connectionUri: connectionUri,
      primaryKey: primaryKey,
      authorizationType: authorizationType,
      authorizationVersion: authorizationVersion,
      xmsVersion: xmsVersion,
      authUtil: _authUtil,
      dateUtil: _dateUtil,
    );

    _databaseDatasource = DatabaseDatasource(
      client: _client,
      connectionUri: connectionUri,
      primaryKey: primaryKey,
      authorizationType: authorizationType,
      authorizationVersion: authorizationVersion,
      xmsVersion: xmsVersion,
      authUtil: _authUtil,
      dateUtil: _dateUtil,
    );

    _documentDatasource = DocumentDatasource(
      client: _client,
      connectionUri: connectionUri,
      primaryKey: primaryKey,
      authorizationType: authorizationType,
      authorizationVersion: authorizationVersion,
      xmsVersion: xmsVersion,
      authUtil: _authUtil,
      dateUtil: _dateUtil,
    );

    _permissionDatasource = PermissionDatasource(
      client: _client,
      connectionUri: connectionUri,
      primaryKey: primaryKey,
      authorizationType: authorizationType,
      authorizationVersion: authorizationVersion,
      xmsVersion: xmsVersion,
      authUtil: _authUtil,
      dateUtil: _dateUtil,
    );

    _triggerDatasource = TriggerDatasource(
      client: _client,
      connectionUri: connectionUri,
      primaryKey: primaryKey,
      authorizationType: authorizationType,
      authorizationVersion: authorizationVersion,
      xmsVersion: xmsVersion,
      authUtil: _authUtil,
      dateUtil: _dateUtil,
    );

    _storedProcedureDatasource = StoredProcedureDatasource(
      client: _client,
      connectionUri: connectionUri,
      primaryKey: primaryKey,
      authorizationType: authorizationType,
      authorizationVersion: authorizationVersion,
      xmsVersion: xmsVersion,
      authUtil: _authUtil,
      dateUtil: _dateUtil,
    );

    _userDatasource = UserDatasource(
      client: _client,
      connectionUri: connectionUri,
      primaryKey: primaryKey,
      authorizationType: authorizationType,
      authorizationVersion: authorizationVersion,
      xmsVersion: xmsVersion,
      authUtil: _authUtil,
      dateUtil: _dateUtil,
    );

    _userDefinedFunctionDatasource = UserDefinedFunctionDatasource(
      client: _client,
      connectionUri: connectionUri,
      primaryKey: primaryKey,
      authorizationType: authorizationType,
      authorizationVersion: authorizationVersion,
      xmsVersion: xmsVersion,
      authUtil: _authUtil,
      dateUtil: _dateUtil,
    );

    collection =
        CollectionRepository(collectionDatasource: _collectionDatasource);
    database = DatabaseRepository(databaseDatasource: _databaseDatasource);
    document = DocumentRepository(documentDatasource: _documentDatasource);
    permission =
        PermissionRepository(permissionDatasource: _permissionDatasource);
    storedProcedure = StoredProcedureRepository(
        storedProcedureDatasource: _storedProcedureDatasource);
    trigger = TriggerRepository(triggerDatasource: _triggerDatasource);
    user = UserRepository(userDatasource: _userDatasource);
    userDefinedFunctions = UserDefinedFunctionRepository(
        userDefinedFunctionDatasource: _userDefinedFunctionDatasource);
  }
}
